clouddrift/wavelet.py

```diff
@@ -0,0 +1,523 @@
+        The time axis of the wavelets must be the last one and matches the length of the time axis of x.
+        The normalization of the wavelets is assumed to be "bandpassed", if not use kwarg norm="energy".
+    boundary : str, optional
+        The boundary condition to be imposed at the edges of the time series.
+        Allowed values are "mirror", "zeros", and "periodic".
+        Default is "mirror".
+    order_axis : int, optional
+        Axis of psi for the order of the wavelets (default is first or 0)
+    f_axis : int, optional
+        Axis of psi for the frequencies of the wavelet (default is second or 1)
+    time_axis : int, optional
+        Axis on which the time is defined for x (default is last, or -1). The time axis of the
+        wavelets must be last.
+
+    Returns
+    -------
+    w : np.ndarray
+        Time-domain wavelet transforms. w shape will be ((series_orders), order, f_axis, time_axis).
+
+    Examples
+    --------
+    To write.
+
+    Raises
+    ------
+    ValueError
+        If the time axis is outside of the valid range ([-1, N-1]).
+        If the shape of time axis is different for input and wavelet.
+        If ``boundary not in ["mirror", "zeros", "periodic"]``.
+
+    See Also
+    --------
+    :func:`morsewave`, `morsefreq`
+    """
+    # time_axis must be in valid range
+    if time_axis < -1 or time_axis > len(x.shape) - 1:
+        raise ValueError(
+            f"time_axis ({time_axis}) is outside of the valid range ([-1,"
+            f" {len(x.shape) - 1}])."
+        )
+    # Positions and time arrays must have the same shape.
+    if x.shape[time_axis] != psi.shape[-1]:
+        raise ValueError("x and psi time axes must have the same length.")
+
+    psi_ = np.moveaxis(psi, [f_axis, order_axis], [-2, -3])
+
+    # initialization: output will be ((x_orders),f_order, f_axis, time_axis)
+    # w = np.tile(
+    #     np.expand_dims(np.zeros_like(x), (-3, -2)),
+    #     (1, np.shape(psi)[-3], np.shape(psi)[-2], 1),
+    # )
+
+    # if x is of dimension 1 we need to expand
+    # otherwise make sure time axis is last
+    if np.ndim(x) < 2:
+        x_ = np.expand_dims(x, axis=0)
+    else:
+        x_ = np.moveaxis(x, time_axis, -1)
+
+    if ~np.all(np.isreal(x)):
+        if norm == "energy":
+            x_ = x_ / np.sqrt(2)
+        elif norm == "bandpass":
+            x_ = x_ / 2
+
+    # to do: add detrending option by default?
+
+    # apply boundary conditions
+    if boundary == "mirror":
+        x_ = np.concatenate((np.flip(x_, axis=-1), x_, np.flip(x_, axis=-1)), axis=-1)
+    elif boundary == "zeros":
+        x_ = np.concatenate((np.zeros_like(x_), x_, np.zeros_like(x_)), axis=-1)
+    elif boundary == "periodic":
+        x_ = np.concatenate((x_, x_, x_), axis=-1)
+    else:
+        raise ValueError("boundary must be one of 'mirror', 'align', or 'zeros'.")
+
+    # danger zone, use different letters from jlab: n is their time length
+    n = np.shape(x)[-1]
```

Comment:
I would suggest using variable names everywhere it's possible. 

For example, `time_length = np.shape(x)[-1)`. Could be good! 

Same below, we have `k, l, w, om`, we could switch to meaningful names too.